src/snowflake/snowpark/modin/plugin/extensions/dataframe_overrides.py

```diff
@@ -459,104 +464,222 @@ def __init__(
-        if distributed_frame is not None:
-            self._query_compiler = distributed_frame._query_compiler
             return
+        # The `columns` parameter is used to select the columns from `data` that will be in the resultant
+        # DataFrame. If a value in `columns` is not present in `data`'s columns, it will be added as a
+        # new column filled with NaN values. These columns are tracked by the `extra_columns` variable.
+        if data.columns is not None and columns is not None:
+            extra_columns = [col for col in columns if col not in data.columns]
+        else:
+            extra_columns = []
+        query_compiler = data._query_compiler.create_qc_with_extra_columns(
+            extra_columns
+        )
 
-        if isinstance(data, native_pd.Index):
-            pass
-        elif is_list_like(data) and not is_dict_like(data):
-            old_dtype = getattr(data, "dtype", None)
-            values = [
-                obj._to_pandas() if isinstance(obj, Series) else obj for obj in data
-            ]
-            if isinstance(data, np.ndarray):
-                data = np.array(values, dtype=old_dtype)
-            else:
-                try:
-                    data = type(data)(values, dtype=old_dtype)
-                except TypeError:
-                    data = values
-        elif is_dict_like(data) and not isinstance(
-            data, (native_pd.Series, Series, native_pd.DataFrame, DataFrame)
-        ):
-            if columns is not None:
-                data = {key: value for key, value in data.items() if key in columns}
-
-            if len(data) and all(isinstance(v, Series) for v in data.values()):
-                from modin.pandas import concat
-
-                new_qc = concat(data.values(), axis=1, keys=data.keys())._query_compiler
-
-                if dtype is not None:
-                    new_qc = new_qc.astype({col: dtype for col in new_qc.columns})
-                if index is not None:
-                    new_qc = new_qc.reindex(
-                        axis=0, labels=try_convert_index_to_native(index)
-                    )
+    else:
+        # CASE V: Non-Snowpark pandas data
+        if not isinstance(
+            data, (native_pd.Series, native_pd.DataFrame, native_pd.Index)
+        ) and is_list_like(data):
+            from snowflake.snowpark.modin.pandas import concat
+
+            if is_dict_like(data):
+                # Setting up keys and values for processing if all the values are Snowpark pandas objects.
                 if columns is not None:
-                    new_qc = new_qc.reindex(
-                        axis=1, labels=try_convert_index_to_native(columns)
-                    )
+                    # Reduce the dictionary to only the relevant columns as the keys.
+                    data = {key: value for key, value in data.items() if key in columns}
+
+                if len(data) and all(
```

Comment:
let's put this into a helper function called like `create_dataframe_from_dict_series`

Reply:
done!